src/app/trips/forms/form-array/form-array.component.ts

```diff
@@ -0,0 +1,86 @@
+import { Component, Input } from "@angular/core";
+import { FormArray, FormControl, FormGroup, FormGroupDirective, Validators } from "@angular/forms";
+import { Options } from "ngx-google-places-autocomplete/objects/options/options";
+
+@Component({
+  selector: 'form-array',
+  templateUrl: 'form-array.component.html',
+  styleUrls: [
+    'form-array.component.css',
+    '../form-field/form-field.component.css',
+    '../trip-form/trip-form.component.css'
+  ]
+})
+
+
+export class FormArrayComponent {
+  form!:FormGroup;
+  array!:FormControl;
+  @Input() hasInvalidStyling?:boolean;
+  @Input() toggleAlert?:boolean;
+  @Input() labelText?:string;
+  @Input() itemText?:string;
+  @Input() $placeholder!:string;
+  @Input() $formArrayName!:string;
+  @Input() $id!:string;
+  @Input() isRequired!:boolean;
+  mouseoverAddValue?:boolean;
+
+  // locally set props
+  alertMessage!:string;
+
+  //for dynamically added destination autocomplete ids
+  autocompleteOptions:Options = new Options ({
+    fields: ["address_components", "name", "formatted_address", "place_id"],
+    types: ["(cities)"]
+  });
+  autocompletePlaceholder!:string;
+
+  constructor(private rootFormGroup:FormGroupDirective) {}
+
+  ngOnInit() {
+    this.form = this.rootFormGroup.control;
+    this.array = this.form.get(this.$formArrayName) as FormControl;
+    this.alertMessage = this.itemText + ' must be valid before adding another ' + this.itemText + '.';
+    this.autocompletePlaceholder = this.$placeholder;
+  }
+
+  // get values array
+  get values(): FormArray{ // capitalized to prevent duplicate identifier
+    return <FormArray>this.form.get(this.$formArrayName);
+  }
+
+  // value methods
+  buildValues():FormControl {
+    if (this.isRequired) {
+      return new FormControl('',Validators.required);
+    }
+    return new FormControl('');
+  }
+  addValue(i: number) {
+    // add value to Values array
+    this.values.push(this.buildValues());
+
+    // set added value as unchangeable and adjust appearance appropriately
+    this.styleInputAsReadonly(this.$formArrayName,i);
+    this.mouseoverAddValue=false; // disables mouseover add button immediately
+  }
+  removeValue(i: number) {
+    this.values.removeAt(i);
+  }
+
+  // make values READONLY when added
+  // fieldName = $formArrayName+'Field'
+  styleInputAsReadonly(fieldName: string, i: number) {
+    // make input READONLY and remove bottom border
+    let fieldInput = document.getElementById(fieldName+i);
```

Comment:
You don't want to use `document` directly. If you must do dom manipulation then you should use [ViewChild](https://angular.io/api/core/ViewChild) but you should try to do this without DOM manipulation. Maybe you could conditionally control a `disabled` flag that is used in the HTML?

Reply:
Additionally, you don't want to add styling in TS. The styling should be done in CSS based on the changes you make to the element. You can do this through pseudo classes like:
```
input:disabled {
    background-color #e4e4f4;
}
```